gax/pagination.h

```diff
@@ -16,60 +16,145 @@
+    iterator end() const { return ElementAccessor{}(raw_page_)->cend(); }
+
+    std::string NextPageToken() const { return raw_page_.next_page_token(); }
+    PageType const& RawPage() const { return raw_page_; }
+
+   private:
+    friend Pages::iterator;
+    PageType raw_page_;
+  };
+
   class iterator {
    public:
     using iterator_category = std::forward_iterator_tag;
-    using value_type = ElementType;
+    using value_type = PageResult;
     using difference_type = std::ptrdiff_t;
-    using pointer = ElementType*;
-    using reference = ElementType&;
+    using pointer = PageResult*;
+    using reference = PageResult&;
 
-    reference operator*() const { return *current_; }
-    pointer operator->() const { return &(*current_); }
+    PageResult const& operator*() const { return page_result_; }
+    PageResult const* operator->() const { return &page_result_; }
     iterator& operator++() {
-      ++current_;
+      // Note: if the rpc fails, the page will be untouched,
+      // i.e. will be have an empty page token and element collection.
+      // This invalidates any iterators on the PageResult.
+      page_result_.raw_page_.Clear();
+      get_next_page_(&page_result_.raw_page_);
+      num_pages_++;
       return *this;
     }
+
+    // Just want to compare against end()
     bool operator==(iterator const& rhs) const {
-      return current_ == rhs.current_;
+      return page_result_.NextPageToken() == rhs.page_result_.NextPageToken() ||
+             num_pages_ == rhs.num_pages_;
     }
     bool operator!=(iterator const& rhs) const { return !(*this == rhs); }
 
-    iterator(FieldIterator current) : current_(std::move(current)) {}
+   private:
+    friend Pages;
+    // Note: don't need to provide a const-lvalue receiving version of the
+    // constructor since iterators always take ownership of a page and pages can
+    // be very expensive to copy if they have a lot of repeated elements.
+    iterator(PageType&& page_result, NextPageRetriever get_next_page,
+             int num_pages)
+        : page_result_(std::move(page_result)),
+          get_next_page_(std::move(get_next_page)),
+          num_pages_(num_pages) {}
 
-    FieldIterator current_;
+    PageResult page_result_;
+    NextPageRetriever get_next_page_;
+    int num_pages_;
   };
-  // Note: Since the constructor will eventually be private (with friend
-  // access), we don't have to define all the variants for references.
-  // We just want to take ownership of a page.
-  PageResult(PageType&& raw_page, ElementAccessor accessor)
-      : raw_page_(std::move(raw_page)), accessor_(accessor) {}
 
-  iterator begin() { return accessor_(raw_page_)->begin(); }
-  iterator end() { return accessor_(raw_page_)->end(); }
+  // Note that a cap of 0 indicates that an iterator will only terminate
+  // when the next page token is empty.
+  Pages(NextPageRetriever get_next_page, int cap)
```

Comment:
What is `cap`? Is that a limit on the number of pages or a limit on the number of results?  I can figure it out, but make it self-describing or document what it is.

Reply:
It's a cap on the number of pages to fetch. I'll update the name and annotate it in method comments.